### PR TITLE
feat(app-delivery) IAM policy for deploy stack

### LINKED ISSUE #1151: Correct docs for app-delivery & improve usability
1. (Found by @moofish32 on Gitter) This is wrong:

```ts
const project = new codebuild.PipelineProject(pipelineStack, 'CodeBuild', {
  /* ... */
});
const synthesizedApp = project.outputArtifact;
```

`Project` doesn't have an `outputArtifact` property, it's a CodePipeline thing. It's also missing a build Pipeline Stage.

So, it should be something like:

```ts
const project = new codebuild.PipelineProject(pipelineStack, 'CodeBuild', {
  /* ... */
});
const buildStage = pipeline.addStage('build');
const buildAction = project.addBuildToPipeline(buildStage, 'CodeBuild');
const synthesizedApp = buildAction.outputArtifact;
```

2. It uses `import cicd = require('@aws-cdk/cicd');`, which doesn't exist anymore (the name was changed).

### PR DESCRIPTION
**Pull request by moofish32**
* The changeset and apply changeset may need specific IAM permissions
 and the user can now customize them via `deployStackAction.role`
 * Document updates for proper build stage configuration
 * Fixes #1151

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license.

### CODE REVIEW

**Comment by eladb** [comment]
@skinny85  can you take a look please?

File: packages/@aws-cdk/app-delivery/lib/pipeline-deploy-stack-action.ts [hunk 1]
```diff
@@ -41,6 +41,29 @@ export interface PipelineDeployStackActionProps {
...
+   * The role to use when creating and executing the ChangeSet.
+   *
+   * @default a new role is created
+   */
+  role?: iam.Role;
+
+  /**
+   * The CloudFormation Capabilities enabled for the ChangeSet.
+   *
+   * @default None
```
[the remark this answers is not in the document]
Comment (rix0rrr) [reply packages/@aws-cdk/app-delivery/lib/pipeline-deploy-stack-action.ts]: I guess this is a request for upstream change then.

@skinny85  ?

[the remark this answers is not in the document]
Comment (RomainMuller) [reply packages/@aws-cdk/app-delivery/lib/pipeline-deploy-stack-action.ts]: The CloudFormation documentation states that `NamedIAM` implies `IAM`, and that you need not specify both.

File: packages/@aws-cdk/app-delivery/README.md [hunk 2]
```diff
@@ -69,26 +73,41 @@ const deployStage = pipeline.addStage('Deploy');
...
 });
+deployServiceBAction.role.addToPolicy(
+ // new iam.PolicyStatement().
+ // ... addAction('actions that you need').
+ // add resource
+);
 ```
 
 #### `buildspec.yml`
+The repository should contain a file at the root level named `buildspec.yml`.
```
Comment (RomainMuller) [inline packages/@aws-cdk/app-delivery/README.md]: That's only true if you don't specify an inline buildspec with your CodeBuild project, so I wouldn't phrase it like so.

File: packages/@aws-cdk/app-delivery/README.md [hunk 3]
```diff
@@ -69,26 +73,41 @@ const deployStage = pipeline.addStage('Deploy');
...
+
 The `PipelineDeployStackAction` expects it's `inputArtifact` to contain the result of synthesizing a CDK App using the
 `cdk synth -o <directory>` command.
 
 For example, a *TypeScript* or *Javascript* CDK App can add the following `buildspec.yml` at the root of the repository
 configured in the `Source` stage:
 
+Example contents of `buildspec.yml` (note: `buildspec.yaml` is not compatible)
```
Comment (RomainMuller) [inline packages/@aws-cdk/app-delivery/README.md]: The precision on `.yaml` is definitely a good one - I'd have moved it in the paragraph just above.

File: packages/@aws-cdk/app-delivery/lib/pipeline-deploy-stack-action.ts [hunk 4]
```diff
@@ -41,6 +41,29 @@ export interface PipelineDeployStackActionProps {
    * @default ``createChangeSetRunOrder + 1``
    */
   executeChangeSetRunOrder?: number;
+
+  /**
+   * The role to use when creating and executing the ChangeSet.
```
Comment (RomainMuller) [inline packages/@aws-cdk/app-delivery/lib/pipeline-deploy-stack-action.ts]: I would copy the comment from the `PipelineCreateReplaceChangeSetActionProps` attribute (same for the other two). I think they're pretty detailed and they avoid insisting so much on the implementation detail here (customer is really just interesting at deploying a stack - the fact there is a ChangeSet being created, then executed, is mostly irrelevant).

File: packages/@aws-cdk/app-delivery/lib/pipeline-deploy-stack-action.ts [hunk 5]
```diff
@@ -69,16 +98,23 @@ export class PipelineDeployStackAction extends cdk.Construct {
       throw new Error(`createChangeSetRunOrder (${createChangeSetRunOrder}) must be < executeChangeSetRunOrder (${executeChangeSetRunOrder})`);
     }
 
-    this.stack = props.stack;
     const changeSetName = props.changeSetName || 'CDK-CodePipeline-ChangeSet';
 
-    new cfn.PipelineCreateReplaceChangeSetAction(this, 'ChangeSet', {
+    this.stack = props.stack;
+
+    const fullPermissions = props.fullPermissions === true;
```
Comment (RomainMuller) [inline packages/@aws-cdk/app-delivery/lib/pipeline-deploy-stack-action.ts]: I don't think this is needed. `fullPermissions` defaults to `false` in `PipelineCreateReplaceChangeSetAction` (and I reckon we're quite unlikely to change this default, as it has pretty drastic security implications).

File: packages/@aws-cdk/app-delivery/package.json [hunk 6]
```diff
@@ -63,6 +65,8 @@
   ],
   "peerDependencies": {
     "@aws-cdk/aws-codepipeline-api": "^0.16.0",
+    "@aws-cdk/aws-cloudformation": "^0.16.0",
```
Comment (RomainMuller) [inline packages/@aws-cdk/app-delivery/package.json]: You'll want to make sure you merge in the `0.17.0` version bump & update those :)

File: packages/@aws-cdk/app-delivery/README.md [hunk 7]
```diff
@@ -52,9 +52,13 @@ const source = new codepipeline.GitHubSourceAction(pipelineStack, 'GitHub', {
   /* ... */
 });
 const project = new codebuild.PipelineProject(pipelineStack, 'CodeBuild', {
-  /* ... */
+  environment: {
+    buildImage: codebuild.LinuxBuildImage.UBUNTU_14_04_NODEJS_10_1_0,
```
Comment (RomainMuller) [inline packages/@aws-cdk/app-delivery/README.md]: I wouldn't specify this in the `README.md`, it could be distracting (or looking like it needs to be that).

File: packages/@aws-cdk/app-delivery/lib/pipeline-deploy-stack-action.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (rix0rrr) [reply packages/@aws-cdk/app-delivery/lib/pipeline-deploy-stack-action.ts]: That future-proofing is exactly what I'm opposing to.

File: packages/@aws-cdk/app-delivery/README.md (hunk 7 above)
Comment (moofish32) [reply packages/@aws-cdk/app-delivery/README.md]: Made a comment instead as an example.

File: packages/@aws-cdk/app-delivery/package.json [hunk 8]
```diff
@@ -46,6 +51,21 @@
     "fast-check": "^1.7.0",
     "pkglint": "^0.17.0"
   },
+  "repository": {
+  },
+  "devDependencies": {
+    "@aws-cdk/aws-codepipeline": "^0.17.0",
+    "@aws-cdk/aws-s3": "^0.17.0",
+    "cdk-build-tools": "^0.17.0",
+    "cdk-integ-tools": "^0.17.0",
```
Comment (skinny85) [inline packages/@aws-cdk/app-delivery/package.json]: Some dependencies are duplicated here (`cdk-build-tools`, `cdk-integ-tools`, `@aws-cdk/aws-s3`, `@aws-cdk/aws-codepipeline`) - I would remove the extra ones  so that the `@aws-cdk` namespace ones are grouped at the beginning.

File: packages/@aws-cdk/app-delivery/lib/pipeline-deploy-stack-action.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (skinny85) [reply packages/@aws-cdk/app-delivery/lib/pipeline-deploy-stack-action.ts]: @moofish32, I believe @rix0rrr means "anonymous IAM permissions" (as opposed to "named IAM permissions").

I agree wholeheartedly with Rico and Romain here. In general, we don't do these kind of future proofing in the CDK - we try to give a nicer API for the common case.

So, I would proceed as follows here:

1. Introduce a new `enum` in this module, with 3 values: `None`, `AnonymousIAM`, `NamedIAM`.
2. Make `capabilities` in `PipelineDeployStackActionProps` be of this new `enum` type (*not* an array), with default `AnonymousIAM`.
3. After this PR, open a separate PR to move this enum to the `aws-cloudformation` package, and change the API of the CFN Actions themselves to use it.

Does this make sense?

File: packages/@aws-cdk/app-delivery/lib/pipeline-deploy-stack-action.ts [hunk 9]
```diff
@@ -69,16 +117,21 @@ export class PipelineDeployStackAction extends cdk.Construct {
       throw new Error(`createChangeSetRunOrder (${createChangeSetRunOrder}) must be < executeChangeSetRunOrder (${executeChangeSetRunOrder})`);
     }
 
-    this.stack = props.stack;
     const changeSetName = props.changeSetName || 'CDK-CodePipeline-ChangeSet';
 
-    new cfn.PipelineCreateReplaceChangeSetAction(this, 'ChangeSet', {
+    this.stack = props.stack;
```
Comment (skinny85) [inline packages/@aws-cdk/app-delivery/lib/pipeline-deploy-stack-action.ts]: Nit: I would not move this line (makes the diff look weird).

File: packages/@aws-cdk/app-delivery/lib/pipeline-deploy-stack-action.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (moofish32) [reply packages/@aws-cdk/app-delivery/lib/pipeline-deploy-stack-action.ts]: If I can do it all in this PR should I? I haven't looked at how difficult this would be plumb.

[the remark this answers is not in the document]
Comment (skinny85) [reply packages/@aws-cdk/app-delivery/lib/pipeline-deploy-stack-action.ts]: Probably easier to leave it out of this PR, as it's a breaking change. But knock yourself out if you want - I don't think it will be super difficult :).

**Comment by rix0rrr** [comment]
I will leave it to @RomainMuller and @skinny85 to merge this PR if they're satisfied.

**Comment by RomainMuller** [comment]
I think we'll be good to merge once @skinny85's suggestion with respects to defining an `enum` locally is implemented.

**Comment by moofish32** [comment]
I have updated the commit (via squash/rebase) and included this breaking change. If we are going to break let's just get on with it.

```
* The changeset and apply changeset can now apply role IAM permissions,
 and CloudFormation Capabilities
 * Updated CloudFormationCapabilities enum to include `None` and renamed `IAM` to `AnonymousIAM`.
 * Document updates for proper build stage configuration
 * Fixes #1151

BREAKING CHANGE: `CloudFormationCapabilities.IAM` renamed to
`CloudFormation.AnonymousIAM` and `PipelineCloudFormationDeployActionProps.capabilities?: CloudFormationCapabilities[]` has been changed to
`PipelineCloudFormationDeployActionProps.capabilities?:
CloudFormationCapabilities` no longer an array.

```

File: packages/@aws-cdk/app-delivery/lib/pipeline-deploy-stack-action.ts [hunk 10]
```diff
@@ -41,6 +41,50 @@ export interface PipelineDeployStackActionProps {
...
+   * capabilities (i.e., able to create all IAM resources).
+   *
+   * This is a shorthand that you can use if you fully trust the templates that
+   * are deployed in this pipeline. If you want more fine-grained permissions,
+   * use `addToRolePolicy` and `capabilities` to control what the CloudFormation
+   * deployment is allowed to do.
+   *
+   * @default false
+   */
+  fullPermissions?: boolean;
```
[the remark this answers is not in the document]
Comment (moofish32) [reply packages/@aws-cdk/app-delivery/lib/pipeline-deploy-stack-action.ts]: I went with `adminPermissions` because when I put in `administrator` I asked my self of what? I thought `adminPermissions` would be more clear, if you are set on `administrator` let me know.

File: packages/@aws-cdk/app-delivery/README.md [hunk 11]
```diff
@@ -69,26 +77,42 @@ const deployStage = pipeline.addStage('Deploy');
...
   stack: serviceStackB,
   inputArtifact: synthesizedApp,
   createChangeSetRunOrder: 998,
 });
+deployServiceBAction.role.addToPolicy(
+ // new iam.PolicyStatement().
+ // ... addAction('actions that you need').
+ // add resource
+);
 ```
```
[the remark this answers is not in the document]
Comment (moofish32) [reply packages/@aws-cdk/app-delivery/README.md]: I didn't add another example but I added text to describe the option; does this work? Or do we want a separate example.

[the remark this answers is not in the document]
Comment (skinny85) [reply packages/@aws-cdk/app-delivery/README.md]: I think, since both examples are the same, I would add a `fullPermissions` example as well.

File: packages/@aws-cdk/app-delivery/README.md [hunk 12]
```diff
@@ -69,26 +77,48 @@ const deployStage = pipeline.addStage('Deploy');
...
   inputArtifact: synthesizedApp,
+  // See the note below for details about this option.
+  adminPermissions: false, 
 });
-new cicd.PipelineDeployStackAction(pipelineStack, 'DeployServiceStackB', {
+
+// Add the necessary permissions for you service deploy action. This role is
+// is passed to CloudFormation and needs the permissions necessary to deploy
+// stack. Alternatively you can enable [Administrator](https://docs.aws.amazon.com/IAM/latest/UserGuide/access_policies_job-functions.html#jf_administrator
+) permissions above,
```
Comment (skinny85) [inline packages/@aws-cdk/app-delivery/README.md]: Huh. This shows up as a new line without comments. Perhaps it's just GitHub's diff displaying it weirdly.

File: packages/@aws-cdk/app-delivery/README.md [hunk 13]
```diff
@@ -69,26 +77,48 @@ const deployStage = pipeline.addStage('Deploy');
...
+The repository can contain a file at the root level named `buildspec.yml`, or
+you can in-line the buildspec. Note that `buildspec.yaml` is not compatible.
+
 The `PipelineDeployStackAction` expects it's `inputArtifact` to contain the result of synthesizing a CDK App using the
 `cdk synth -o <directory>` command.
 
 For example, a *TypeScript* or *Javascript* CDK App can add the following `buildspec.yml` at the root of the repository
 configured in the `Source` stage:
 
+Example contents of `buildspec.yml`.
```
Comment (skinny85) [inline packages/@aws-cdk/app-delivery/README.md]: Example contents of `buildspec.yml`:
// empty line here
```yml
version: 0.2
phases:
```
...

File: packages/@aws-cdk/app-delivery/lib/pipeline-deploy-stack-action.ts [hunk 14]
```diff
@@ -41,6 +41,47 @@ export interface PipelineDeployStackActionProps {
...
+   *
+   * @default A fresh role with admin or no permissions (depending on the value of `adminPermissions`).
+   */
+  role?: iam.Role;
+
+  /**
+   * Acknowledge certain changes made as part of deployment
+   *
+   * For stacks that contain certain resources, explicit acknowledgement that AWS CloudFormation
+   * might create or update those resources. For example, you must specify CAPABILITY_IAM if your
```
Comment (skinny85) [inline packages/@aws-cdk/app-delivery/lib/pipeline-deploy-stack-action.ts]: I believe `CAPABILITY_IAM` is not the correct constant name anymore?

File: packages/@aws-cdk/app-delivery/test/test.pipeline-deploy-stack-action.ts [hunk 15]
```diff
@@ -51,14 +62,170 @@ export = nodeunit.testCase({
...
+    });
+    // we might need to add permissions
+    deployAction.addToRolePolicy( new iam.PolicyStatement().
+      addAction('ec2:AuthorizeSecurityGroupEgress').
+      addAction('ec2:AuthorizeSecurityGroupIngress').
+      addAction('ec2:DeleteSecurityGroup').
+      addAction('ec2:DescribeSecurityGroups').
+      addAction('ec2:CreateSecurityGroup').
+      addAction('ec2:RevokeSecurityGroupEgress').
+      addAction('ec2:RevokeSecurityGroupIngress').
```
Comment (skinny85) [inline packages/@aws-cdk/app-delivery/test/test.pipeline-deploy-stack-action.ts]: Feel free to use the `addActions` method here, that takes a varargs argument, will be more concise.

File: packages/@aws-cdk/aws-cloudformation/lib/pipeline-actions.ts [hunk 16]
```diff
@@ -494,3 +502,16 @@ function stackArnFromProps(props: { stackName: string, region?: string }): strin
...
+    // admin true default capability to NamedIAM
+    return CloudFormationCapabilities.NamedIAM;
+  } else if (capabilities === undefined) {
+    // else capabilities are undefined set AnonymousIAM
+    return CloudFormationCapabilities.AnonymousIAM;
+  } else {
+    // else capabilities are defined use them
+    return capabilities;
+  }
+}
```
Comment (skinny85) [inline packages/@aws-cdk/aws-cloudformation/lib/pipeline-actions.ts]: Sorry, maybe I wasn't clear before.

The capabilities should default to `AnonymousIAM` only in the `app-delivery` construct.

In the CFN Actions, the default should be `undefined` (that is, none), like it is now.

File: packages/@aws-cdk/aws-cloudformation/lib/pipeline-actions.ts [hunk 17]
```diff
@@ -151,10 +151,8 @@ export interface PipelineCloudFormationDeployActionProps extends PipelineCloudFo
    * are deployed in this pipeline. If you want more fine-grained permissions,
    * use `addToRolePolicy` and `capabilities` to control what the CloudFormation
    * deployment is allowed to do.
-   *
-   * @default false
    */
-  fullPermissions?: boolean;
+  adminPermissions: boolean;
```
Comment (skinny85) [inline packages/@aws-cdk/aws-cloudformation/lib/pipeline-actions.ts]: Why did we change this to required? IMO, it should still default to `false`.

Comment (skinny85) [reply packages/@aws-cdk/aws-cloudformation/lib/pipeline-actions.ts]: Ah, I see that was @eladb's comment.

Feel free to leave it required then.

File: packages/@aws-cdk/app-delivery/test/test.pipeline-deploy-stack-action.ts (hunk 15 above)
Comment (moofish32) [reply packages/@aws-cdk/app-delivery/test/test.pipeline-deploy-stack-action.ts]: OH... I didn't see there was one!

File: packages/@aws-cdk/aws-cloudformation/lib/pipeline-actions.ts (hunk 16 above)
Comment (moofish32) [reply packages/@aws-cdk/aws-cloudformation/lib/pipeline-actions.ts]: I wondered about this change. With `adminPermissions` being required it seems a bit odd this default is allow IAM, but it's a pretty heavy coin flip considering it's rare not to deploy a role with a service.

**Comment by moofish32** [comment]
Thanks for the conflict resolution @RomainMuller. @skinny85 did I get all the fixes in here?

File: packages/@aws-cdk/app-delivery/README.md [hunk 18]
```diff
@@ -69,26 +77,44 @@ const deployStage = pipeline.addStage('Deploy');
...
+
+// Add the necessary permissions for you service deploy action. This role is
+// is passed to CloudFormation and needs the permissions necessary to deploy
+// stack. Alternatively you can enable [Administrator](https://docs.aws.amazon.com/IAM/latest/UserGuide/access_policies_job-functions.html#jf_administrator) permissions above, 
+// users should understand the privileged nature of this role.
+deployServiceAAction.addToRolePolicy(
+ // new iam.PolicyStatement().
+ // ... addAction('actions that you need').
+ // add resource
+);
```
Comment (skinny85) [inline packages/@aws-cdk/app-delivery/README.md]: I wouldn't comment out the whole thing, but instead just the usecase-specific part:

```ts
deployServiceAAction.addToRolePolicy(
  new iam.PolicyStatement()
    .addAction('service:SomeAction')
    .addResource(myResource.myResourceArn)
    // add more Action(s) and/or Resource(s) here, as needed
);
```

File: packages/@aws-cdk/app-delivery/README.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (skinny85) [reply packages/@aws-cdk/app-delivery/README.md]: Yep, that's what I meant 👍🏻

File: packages/@aws-cdk/app-delivery/README.md (hunk 13 above)
Comment (skinny85) [reply packages/@aws-cdk/app-delivery/README.md]: I think this now clashes with the sentence above? "For example, a *TypeScript* or *Javascript* CDK App can add the following `buildspec.yml` at the root of the repository configured in the `Source` stage:". Perhaps it always did, I just never noticed ;p.

File: packages/@aws-cdk/aws-cloudformation/lib/pipeline-actions.ts [hunk 19]
```diff
@@ -118,24 +118,24 @@ export interface PipelineCloudFormationDeployActionProps extends PipelineCloudFo
...
    * For stacks that contain certain resources, explicit acknowledgement that AWS CloudFormation
-   * might create or update those resources. For example, you must specify CAPABILITY_IAM if your
-   * stack template contains AWS Identity and Access Management (IAM) resources. For more
-   * information, see [Acknowledging IAM Resources in AWS CloudFormation Templates](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/using-iam-template.html#using-iam-capabilities).
-   *
-   * @default No capabitilities passed, unless `fullPermissions` is true
+   * might create or update those resources. For example, you must specify `AnonymousIAM` or `NamedIAM`
+   * if your stack template contains AWS Identity and Access Management (IAM) resources. For more
+   * information
+   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/using-iam-template.html#using-iam-capabilities
```
Comment (skinny85) [inline packages/@aws-cdk/aws-cloudformation/lib/pipeline-actions.ts]: I don't think this will render well (the sentence will be cut off, and the @see will be below the text paragraph. Feel free to use Markdown links here: `For more information, [see the AWS documentation](https://docs.aws.amazon.com/somepath)`.

File: packages/@aws-cdk/aws-cloudformation/lib/pipeline-actions.ts [hunk 20]
```diff
@@ -366,8 +364,17 @@ export enum CloudFormationCapabilities {
...
+  NamedIAM = 'CAPABILITY_NAMED_IAM',
+
+  /**
+   * No IAM Capabilities
+   *
+   * Pass this capability if you wish to block the creation IAM resources.
+   * @link https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/using-iam-template.html#using-iam-capabilities
    */
-  NamedIAM = 'CAPABILITY_NAMED_IAM'
+  None = '',
```
Comment (skinny85) [inline packages/@aws-cdk/aws-cloudformation/lib/pipeline-actions.ts]: I would actually make `None` the first enum value (I would sort them in the order of increasing capabilities).

File: packages/@aws-cdk/aws-codepipeline/test/integ.cfn-template-from-repo.lit.expected.json [hunk 21]
```diff
@@ -322,4 +322,4 @@
       }
     }
   }
-}
+}
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codepipeline/test/integ.cfn-template-from-repo.lit.expected.json]: Can you try to revert this change? I think the integ tests will pass without this ordering change.

File: packages/@aws-cdk/aws-codepipeline/test/integ.lambda-pipeline.expected.json [hunk 22]
```diff
@@ -181,7 +175,13 @@
...
             "Name": "Lambda"
           }
-        ]
+        ],
+        "ArtifactStore": {
+          "Location": {
+            "Ref": "PipelineArtifactsBucket22248F97"
+          },
+          "Type": "S3"
+        }
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codepipeline/test/integ.lambda-pipeline.expected.json]: Same comment here (try to revert this change and see if the compare tests still pass, they should).

File: packages/@aws-cdk/aws-codepipeline/test/integ.pipeline-cfn.expected.json [hunk 23]
```diff
@@ -254,4 +254,4 @@
       }
     }
   }
-}
+}
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codepipeline/test/integ.pipeline-cfn.expected.json]: Same comment here as in the other 2 expected JSON files.

File: packages/@aws-cdk/aws-codepipeline/test/integ.pipeline-code-build-multiple-inputs-outputs.expected.json [hunk 24]
```diff
@@ -355,7 +349,13 @@
...
             "Name": "Build"
           }
-        ]
+        ],
+        "ArtifactStore": {
+          "Location": {
+            "Ref": "MyBucketF68F3FF0"
+          },
+          "Type": "S3"
+        }
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codepipeline/test/integ.pipeline-code-build-multiple-inputs-outputs.expected.json]: Same comment here as in the other 3 expected JSON files.

File: packages/@aws-cdk/aws-codepipeline/test/integ.pipeline-code-commit-build.expected.json [hunk 25]
```diff
@@ -219,7 +213,13 @@
...
             "Name": "build"
           }
-        ]
+        ],
+        "ArtifactStore": {
+          "Location": {
+            "Ref": "PipelineArtifactsBucket22248F97"
+          },
+          "Type": "S3"
+        }
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codepipeline/test/integ.pipeline-code-commit-build.expected.json]: Same comment here as in the other 4 expected JSON files.

File: packages/@aws-cdk/aws-codepipeline/test/integ.pipeline-code-commit.expected.json [hunk 26]
```diff
@@ -221,7 +215,13 @@
...
             "Name": "build"
           }
-        ]
+        ],
+        "ArtifactStore": {
+          "Location": {
+            "Ref": "PipelineArtifactsBucket22248F97"
+          },
+          "Type": "S3"
+        }
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codepipeline/test/integ.pipeline-code-commit.expected.json]: Same comment here as in the other 5 expected JSON files.

File: packages/@aws-cdk/aws-codepipeline/test/integ.pipeline-code-deploy.expected.json [hunk 27]
```diff
@@ -328,7 +322,13 @@
...
             "Name": "Deploy"
           }
-        ]
+        ],
+        "ArtifactStore": {
+          "Location": {
+            "Ref": "CodeDeployPipelineIntegTest9F618D61"
+          },
+          "Type": "S3"
+        }
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codepipeline/test/integ.pipeline-code-deploy.expected.json]: Same comment here as in the other 6 expected JSON files.

File: packages/@aws-cdk/aws-codepipeline/test/integ.pipeline-events.expected.json [hunk 28]
```diff
@@ -103,12 +103,6 @@
     "MyPipelineAED38ECF": {
       "Type": "AWS::CodePipeline::Pipeline",
       "Properties": {
-        "ArtifactStore": {
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codepipeline/test/integ.pipeline-events.expected.json]: Same comment here as in the other 7 expected JSON files.

File: packages/@aws-cdk/aws-cloudformation/lib/pipeline-actions.ts [hunk 29]
```diff
@@ -198,12 +196,11 @@ export abstract class PipelineCloudFormationDeployAction extends PipelineCloudFo
...
   constructor(parent: cdk.Construct, id: string, props: PipelineCloudFormationDeployActionProps, configuration: any) {
-    const capabilities = props.fullPermissions && props.capabilities === undefined ? [CloudFormationCapabilities.NamedIAM] : props.capabilities;
-
+    const capabilities = props.adminPermissions && props.capabilities === undefined ? CloudFormationCapabilities.NamedIAM : props.capabilities;
     super(parent, id, props, {
       ...configuration,
-      // This must be a string, so flatten the list to a comma-separated string.
-      Capabilities: (capabilities && capabilities.join(',')) || undefined,
+      // None evaluates to empty string which is falsey and results in undefined
+      Capabilities: capabilities && capabilities.toString() || undefined,
```
Comment (skinny85) [inline packages/@aws-cdk/aws-cloudformation/lib/pipeline-actions.ts]: Oooh, you're relying on the fact that `&&` binds stronger than `||` - that is pure evil :). Mind adding some parenthesis around the 'and' expression?

**Comment by moofish32** [comment]
alright @skinny85 one more pass when ever you are back from the holiday.

File: packages/@aws-cdk/aws-cloudformation/lib/pipeline-actions.ts (hunk 19 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-cloudformation/lib/pipeline-actions.ts]: Ok, this way is fine 🙂
